File: media-server/src/services/audio/YtDlpAudioPlayer.ts
import { EventEmitter } from 'events';
import { IAudioPlayer, IPlaybackStatus } from './AudioPlayer';
import Speaker from 'speaker';
import { spawn } from 'child_process';
import { join } from 'path';
import { createHash } from 'crypto';
import { existsSync, mkdirSync, createReadStream, rmSync } from 'fs';

interface TrackInfo {
  duration: number;
  title: string;
}

export class YtDlpAudioPlayer extends EventEmitter implements IAudioPlayer {
  private speaker: Speaker | null = null;
  private ffmpeg: any = null;
  private ytDlp: any = null;
  private status: IPlaybackStatus = {
    isPlaying: false
  };
  private startTime: number = 0;
  private readonly cacheDir: string;

  constructor() {
    super();
    // Create cache directory in user's home directory
    this.cacheDir = join(process.env.HOME || process.env.USERPROFILE || '.', '.audio-cache');
    if (!existsSync(this.cacheDir)) {
      mkdirSync(this.cacheDir, { recursive: true });
    }
  }

  private getCacheFilePath(url: string): string {
    const hash = createHash('md5').update(url).digest('hex');
    return join(this.cacheDir, `${hash}.opus`); // Using .opus since we prefer opus codec
  }

  private async playLoadingSound(): Promise<void> {
    return new Promise((resolve, reject) => {
      try {
        // Create ffmpeg process for the sound effect
        const ffmpeg = spawn('ffmpeg', [
          '-i', join(__dirname, '../../../assets/loading-1.wav'),
          '-loglevel', 'error',    // Only show actual errors
          '-hide_banner',          // Hide FFmpeg compilation info
          '-f', 's16le',
          '-ar', '48000',
          '-ac', '2',
          'pipe:1'
        ]);

        // Create temporary speaker for the sound effect
        const loadingSpeaker = new Speaker({
          channels: 2,
          bitDepth: 16,
          sampleRate: 48000
        });

        // Handle completion
        loadingSpeaker.on('finish', () => {
          resolve();
        });

        // Handle only real errors
        ffmpeg.stderr.on('data', (data: Buffer) => {
          const error = data.toString().trim();
          if (error) {  // Only log if there's an actual error
            console.error('Loading sound effect error:', error);
          }
        });

        // Pipe the audio
        ffmpeg.stdout.pipe(loadingSpeaker);

      } catch (error) {
        console.error('Failed to play loading sound:', error);
        resolve(); // Resolve anyway to continue with main playback
      }
    });
  }

  async play(url: string): Promise<void> {
    try {
      // Ensure previous playback is fully stopped before starting new one
      await this.stop();
      
      // Add a small delay to ensure cleanup is complete
      await new Promise(resolve => setTimeout(resolve, 100));

      const cacheFile = this.getCacheFilePath(url);
      let trackInfo: TrackInfo;

      if (!existsSync(cacheFile)) {
        this.playLoadingSound().catch(console.error);
        // Get info and download if not cached
        trackInfo = await this.getTrackInfo(url);
        
        console.log('\n▶️ Downloading:', trackInfo.title);
        console.log(`⏱️ Duration: ${Math.floor(trackInfo.duration / 60)}:${(trackInfo.duration % 60).toString().padStart(2, '0')}`);
        
        await this.downloadToCache(url, cacheFile);
      } else {
        // Get duration of cached file
        const duration = await this.getCachedFileDuration(cacheFile);
        // Try to get title from metadata, fallback to URL
        const title = await this.getCachedFileTitle(cacheFile) || url;
        trackInfo = { duration, title };
        
        console.log('\n▶️ Playing from cache:', trackInfo.title);
        console.log(`⏱️ Duration: ${Math.floor(trackInfo.duration / 60)}:${(trackInfo.duration % 60).toString().padStart(2, '0')}`);
      }

      // Play the cached file
      await this.playFromCache(cacheFile, trackInfo);

    } catch (error) {
      console.error('Error in YtDlpAudioPlayer:', error);
      await this.stop();
      throw error;
    }
  }

  private async getTrackInfo(url: string): Promise<TrackInfo> {
    const infoProcess = spawn('yt-dlp', [
      '--get-title',       // This should come first
      '--get-duration',
      '--no-playlist',
      '--no-check-formats',
      '--no-warnings',
      '--extract-audio',   // Add this to ensure we get audio info
      url
    ]);

    return new Promise<TrackInfo>((resolve, reject) => {
      let output = '';
      infoProcess.stdout.on('data', (data: Buffer) => {
        output += data.toString();
      });
      
      infoProcess.on('close', (code) => {
        if (code !== 0) {
          reject(new Error('Failed to get track info'));
          return;
        }
        const lines = output.trim().split('\n');
        if (lines.length < 2) {
          reject(new Error('Invalid yt-dlp output format'));
          return;
        }
        
        const title = lines[0].trim();
        const durationStr = lines[1].trim();
        
        console.log('Debug - Title:', title); // Debug log
        console.log('Debug - Duration:', durationStr); // Debug log
        
        const parts = durationStr.split(':').map(Number);
        let seconds = 0;
        if (parts.length === 3) { // HH:MM:SS
          seconds = parts[0] * 3600 + parts[1] * 60 + parts[2];
        } else if (parts.length === 2) { // MM:SS
          seconds = parts[0] * 60 + parts[1];
        } else { // SS
          seconds = parts[0];
        }
        resolve({ duration: seconds, title });
      });

      infoProcess.stderr.on('data', (data: Buffer) => {
        console.error('yt-dlp error:', data.toString());
      });

      infoProcess.on('error', reject);
    });
  }

  private async getCachedFileTitle(filePath: string): Promise<string | null> {
    return new Promise((resolve) => {
      const ffprobe = spawn('ffprobe', [
        '-v', 'error',
        '-select_streams', 'a:0', // Select audio stream
        '-show_entries', 'format_tags=title,TITLE:stream_tags=title,TITLE', // Check both format and stream tags
        '-of', 'json=c=1',  // Output in JSON format for easier parsing
        filePath
      ]);

      let output = '';
      ffprobe.stdout.on('data', (data) => {
        output += data.toString();
      });

      ffprobe.on('close', (code) => {
        if (code === 0 && output.trim()) {
          try {
            const metadata = JSON.parse(output);
            // Try to find title in different possible locations
            const title = metadata.format?.tags?.title || 
                         metadata.format?.tags?.TITLE ||
                         metadata.streams?.[0]?.tags?.title ||
                         metadata.streams?.[0]?.tags?.TITLE;
            
            if (title) {
              resolve(title);
            } else {
              resolve(null);
            }
          } catch (e) {
            console.error('Error parsing ffprobe output:', e);
            resolve(null);
          }
        } else {
          resolve(null);
        }
      });

      ffprobe.on('error', () => resolve(null));
    });
  }

  private async downloadToCache(url: string, cacheFile: string): Promise<void> {
    return new Promise<void>((resolve, reject) => {
      const ytDlp = spawn('yt-dlp', [
        '-f', 'bestaudio[acodec=opus]/bestaudio',
        '--no-playlist',
        '--no-check-formats',
        '--no-warnings',
        '--force-ipv4',
        '--no-part',
        '--no-cache-dir',
        '--prefer-ffmpeg',
        '--add-metadata',        // Add this to include metadata
        '--embed-metadata',      // And this to embed it in the file
        '--embed-info-json',     // Embed all info in the file
        '--write-info-json',     // Also write info to separate file
        '-o', cacheFile,
        url
      ]);

      ytDlp.stderr.on('data', (data: Buffer) => {
        const error = data.toString().trim();
        if (error && !error.includes('Download completed')) {
          console.error('yt-dlp error:', error);
        }
      });

      ytDlp.on('close', (code) => {
        if (code === 0) {
          resolve();
        } else {
          reject(new Error(`yt-dlp exited with code ${code}`));
        }
      });

      ytDlp.on('error', reject);
    });
  }

  private async getCachedFileDuration(filePath: string): Promise<number> {
    return new Promise<number>((resolve, reject) => {
      const ffprobe = spawn('ffprobe', [
        '-v', 'error',
        '-show_entries', 'format=duration',
        '-of', 'default=noprint_wrappers=1:nokey=1',
        filePath
      ]);

      let output = '';
      ffprobe.stdout.on('data', (data) => {
        output += data.toString();
      });

      ffprobe.on('close', (code) => {
        if (code === 0) {
          resolve(parseFloat(output.trim()));
        } else {
          reject(new Error('Failed to get duration'));
        }
      });

      ffprobe.on('error', reject);
    });
  }

  private async playFromCache(filePath: string, trackInfo: TrackInfo): Promise<void> {
    // Create ffmpeg process to decode the cached file
    this.ffmpeg = spawn('ffmpeg', [
      '-i', filePath,
      '-ar', '48000',
      '-ac', '2',
      '-acodec', 'pcm_s16le',
      '-f', 's16le',
      '-bufsize', '8192',
      '-loglevel', 'error',
      'pipe:1'
    ]);

    // Create speaker
    this.speaker = new Speaker({
      channels: 2,
      bitDepth: 16,
      sampleRate: 48000
    });

    // Set up error handling
    this.ffmpeg.stdout.on('error', (error: Error) => {
      if (!error.message.includes('EPIPE')) {
        console.error('ffmpeg output stream error:', error);
        this.emit('error', error);
      }
    });

    // Store the finish handler so we can remove it later
    const finishHandler = () => {
      if (this.speaker === speaker) { // Only update status if this is still the current speaker
        this.status.isPlaying = false;
        console.log('Debug - Stopped from finish');
        this.emit('stopped');
      }
    };

    const speaker = this.speaker; // Store reference to current speaker
    speaker.on('finish', finishHandler);

    // Connect the pipeline
    this.ffmpeg.stdout.pipe(this.speaker);

    // Update status with title
    this.status = {
      isPlaying: true,
      currentUrl: filePath,
      duration: trackInfo.duration,
      title: trackInfo.title
    };
    this.startTime = Date.now();

    // Handle completion
    this.speaker.on('finish', () => {
      this.status.isPlaying = false;
      console.log('Debug - Stopped from finish');
      this.emit('stopped');
    });

    // Start progress updates
    this.startProgressUpdates();
  }

  private startProgressUpdates() {
    let lastLines = 0; // Keep track of how many lines we printed
    const updateInterval = setInterval(() => {
      if (this.status.isPlaying && this.status.duration) {
        const elapsed = (Date.now() - this.startTime) / 1000;
        this.status.progress = Math.min(elapsed / this.status.duration, 1);
        
        // Create a progress bar
        const width = 30;
        const complete = Math.floor(this.status.progress * width);
        const incomplete = width - complete;
        const progressBar = '█'.repeat(complete) + '▒'.repeat(incomplete);
        
        // Calculate time
        const currentTime = Math.floor(elapsed);
        const totalTime = Math.floor(this.status.duration);
        const currentMin = Math.floor(currentTime / 60);
        const currentSec = (currentTime % 60).toString().padStart(2, '0');
        const totalMin = Math.floor(totalTime / 60);
        const totalSec = (totalTime % 60).toString().padStart(2, '0');
        
        // Clear previous lines
        if (lastLines > 0) {
          process.stdout.write(`\x1b[${lastLines}A\x1b[0J`);
        }
        
        // Show new status
        console.log(`⏯️  ${this.status.title}`);
        console.log(`   ${progressBar} ${currentMin}:${currentSec}/${totalMin}:${totalSec}`);
        
        lastLines = 2; // We printed 2 lines
        
        if (this.status.progress >= 1) {
          console.log('✅ Playback complete');
          clearInterval(updateInterval);
        }
      } else {
        clearInterval(updateInterval);
      }
    }, 1000);
  }

  async stop(): Promise<void> {
    try {
      if (!this.speaker && !this.ffmpeg && !this.ytDlp) {
        return; // Nothing to stop
      }

      return new Promise<void>((resolve) => {
        // First, end input streams
        if (this.ytDlp) {
          this.ytDlp.stdout.unpipe();
          this.ytDlp.kill('SIGKILL');
          this.ytDlp = null;
        }

        // Wait for ffmpeg to finish processing
        if (this.ffmpeg) {
          this.ffmpeg.stdin.end();
          this.ffmpeg.stdout.unpipe();
          this.ffmpeg.kill('SIGKILL');
          this.ffmpeg = null;
        }

        // Clean up speaker and wait for it to finish
        if (this.speaker) {
          this.speaker.on('close', () => {
            this.speaker = null;
            this.status = { isPlaying: false };
            this.emit('stopped');
            resolve();
          });
          this.speaker.end();
        } else {
          resolve();
        }
      });

    } catch (error) {
      console.error('Error during stop:', error);
      // Force cleanup in case of error
      this.ytDlp = null;
      this.ffmpeg = null;
      this.speaker = null;
      this.status.isPlaying = false;
      throw error;
    }
  }

  async pause(): Promise<void> {
    if (this.speaker) {
      // @ts-ignore - Speaker does have pause() but it's not in the types
      this.speaker.pause();
      this.status.isPlaying = false;
      this.emit('paused');
    }
  }

  async resume(): Promise<void> {
    if (this.speaker) {
      // @ts-ignore - Speaker does have resume() but it's not in the types
      this.speaker.resume();
      this.status.isPlaying = true;
      this.emit('resumed');
    }
  }

  getStatus(): IPlaybackStatus {
    return { ...this.status };
  }

  public clearCache(): void {
    try {
      rmSync(this.cacheDir, { recursive: true, force: true });
      mkdirSync(this.cacheDir, { recursive: true });
      console.log('Cache cleared successfully');
    } catch (error) {
      console.error('Error clearing cache:', error);
    }
  }
} 

File: media-server-py/src/services/audio/ytdlp_audio_player.py
from typing import Dict, Callable, List, Optional
import asyncio
import os
import json
import hashlib
from pathlib import Path
import vlc
import yt_dlp

# Import the EinkDisplayManager
from ..display.eink_manager import EinkDisplayManager

class YtDlpAudioPlayer:
    def __init__(self, use_eink_display=True):
        # Debug audio devices
        print("Available audio devices:")
        instance = vlc.Instance()
        mods = instance.audio_output_enumerate_devices()
        if mods:
            for m in mods:
                print(f"Device: {m}")
        else:
            print("No audio devices found!")
        
        # Initialize VLC with ALSA
        self._vlc_instance = vlc.Instance('--no-xlib', '--aout=alsa', '--alsa-audio-device=hw:1,0')
        self._player = self._vlc_instance.media_player_new()
        self._current_media = None
        self._status = {"is_playing": False}
        self._start_time = 0
        
        # Set up cache directory
        home = os.environ.get('HOME', os.environ.get('USERPROFILE', '.'))
        self._cache_dir = Path(home) / '.audio-cache'
        self._cache_dir.mkdir(parents=True, exist_ok=True)
        
        # Event handlers
        self._event_handlers: Dict[str, List[Callable]] = {
            'error': [],
            'stopped': []
        }
        
        # Set up VLC event manager
        self._event_manager = self._player.event_manager()
        self._event_manager.event_attach(vlc.EventType.MediaPlayerEndReached, 
                                       self._on_playback_finished)

        # Initialize e-ink display
        self._use_eink_display = use_eink_display
        if self._use_eink_display:
            try:
                self._display = EinkDisplayManager()
            except Exception as e:
                print(f"Error initializing e-ink display: {e}")
                self._use_eink_display = False

    def on(self, event: str, callback: Callable) -> None:
        """Register an event handler"""
        if event in self._event_handlers:
            self._event_handlers[event].append(callback)
    
    def _emit(self, event: str, *args) -> None:
        """Emit an event to all registered handlers"""
        if event in self._event_handlers:
            for handler in self._event_handlers[event]:
                handler(*args)

    def _on_playback_finished(self, event) -> None:
        """Handle playback finished event"""
        self._status["is_playing"] = False
        print('Debug - Stopped from finish')
        # Update the display to standby mode
        if self._use_eink_display:
            self._display.show_standby()
        self._emit('stopped')

    def _get_cache_file_path(self, url: str) -> Path:
        """Generate cache file path from URL"""
        url_hash = hashlib.md5(url.encode()).hexdigest()
        return self._cache_dir / f"{url_hash}.opus"
        
    def _get_metadata_file_path(self, url: str) -> Path:
        """Generate metadata file path from URL"""
        url_hash = hashlib.md5(url.encode()).hexdigest()
        return self._cache_dir / f"{url_hash}.json"
        
    async def _save_track_metadata(self, url: str, metadata: Dict) -> None:
        """Save track metadata to cache"""
        metadata_file = self._get_metadata_file_path(url)
        try:
            with open(metadata_file, 'w') as f:
                json.dump(metadata, f)
        except Exception as e:
            print(f"Error saving metadata: {e}")
            
    async def _load_track_metadata(self, url: str) -> Optional[Dict]:
        """Load track metadata from cache if available"""
        metadata_file = self._get_metadata_file_path(url)
        if metadata_file.exists():
            try:
                with open(metadata_file, 'r') as f:
                    return json.load(f)
            except Exception as e:
                print(f"Error loading metadata: {e}")
        return None

    async def _get_track_info(self, url: str) -> Dict:
        """Get track information using yt-dlp"""
        ydl_opts = {
            'quiet': True,
            'no_warnings': True,
            'extract_flat': True,
            'format': 'bestaudio'
        }
        
        try:
            with yt_dlp.YoutubeDL(ydl_opts) as ydl:
                info = await asyncio.to_thread(ydl.extract_info, url, download=False)
                return {
                    'duration': info.get('duration', 0),
                    'title': info.get('title', url)
                }
        except Exception as e:
            print(f"Error getting track info: {e}")
            raise

    async def play(self, url: str) -> None:
        """Play audio from URL"""
        try:
            # Stop any current playback
            await self.stop()
            await asyncio.sleep(0.1)  # Small delay for cleanup

            # Show loading screen on e-ink display
            if self._use_eink_display:
                self._display.show_loading("Downloading...")

            print(f"Checking cache for {url}")
            cache_file = self._get_cache_file_path(url)
            
            print(f"Cache file: {cache_file}")
            if not cache_file.exists():
                print('\n▶️ Downloading...')
                track_info = await self._get_track_info(url)
                await self._download_to_cache(url, cache_file)
                # Save metadata after download
                await self._save_track_metadata(url, track_info)
            else:
                print(f"Cache file found for {url}")
                # Try to load metadata from cache
                track_info = await self._load_track_metadata(url)
                if track_info is None:
                    # Fall back to online lookup if metadata not cached
                    print("Metadata not cached, fetching online...")
                    track_info = await self._get_track_info(url)
                    await self._save_track_metadata(url, track_info)
                print('\n▶️ Playing from cache')

            print(f"▶️ Playing: {track_info['title']}")
            duration_min = track_info['duration'] // 60
            duration_sec = track_info['duration'] % 60
            print(f"⏱️ Duration: {duration_min}:{duration_sec:02d}")

            # Update e-ink display with initial information
            if self._use_eink_display:
                current_time = "0:00"
                total_time = f"{duration_min}:{duration_sec:02d}"
                self._display.show_playback(track_info['title'], current_time, total_time, 0)

            # Play the cached file
            self._current_media = self._vlc_instance.media_new(str(cache_file))
            self._player.set_media(self._current_media)
            self._player.play()
            
            self._status = {
                "is_playing": True,
                "current_url": str(cache_file),
                "duration": track_info['duration'],
                "title": track_info['title']
            }
            # Get current time safely
            self._start_time = asyncio.get_running_loop().time()
            
            # Start progress updates
            asyncio.create_task(self._update_progress())

        except Exception as error:
            print(f"Error in play: {error}")
            await self.stop()
            raise

    async def _download_to_cache(self, url: str, cache_file: Path) -> None:
        """Download audio to cache using yt-dlp"""
        ydl_opts = {
            'format': 'bestaudio[acodec=opus]/bestaudio',
            'outtmpl': str(cache_file),
            'quiet': True,
            'no_warnings': True,
            'extract_audio': True,
            'audio_format': 'opus'
        }
        
        try:
            with yt_dlp.YoutubeDL(ydl_opts) as ydl:
                await asyncio.to_thread(ydl.download, [url])
            await self._save_track_metadata(url, await self._get_track_info(url))
        except Exception as e:
            print(f"Download error: {e}")
            raise

    async def _update_progress(self) -> None:
        """Update playback progress"""
        last_lines = 0
        while self._status.get("is_playing", False):
            if "duration" in self._status:
                # Get current time safely
                elapsed = asyncio.get_running_loop().time() - self._start_time
                progress = min(elapsed / self._status["duration"], 1)
                
                # Create progress bar
                width = 30
                complete = int(progress * width)
                incomplete = width - complete
                progress_bar = '█' * complete + '▒' * incomplete
                
                # Calculate times
                current_time = int(elapsed)
                total_time = int(self._status["duration"])
                current_min, current_sec = divmod(current_time, 60)
                total_min, total_sec = divmod(total_time, 60)
                
                current_time_str = f"{current_min}:{current_sec:02d}"
                total_time_str = f"{total_min}:{total_sec:02d}"
                
                # Clear previous lines and redraw progress in terminal
                if last_lines > 0:
                    # Move cursor up and clear lines
                    print(f"\033[{last_lines}A\033[J", end='')
                
                # Show new status in terminal
                print(f"⏯️  {self._status.get('title', 'Unknown')}")
                print(f"   {progress_bar} {current_time_str}/{total_time_str}")
                
                last_lines = 2
                
                # Update e-ink display
                if self._use_eink_display:
                    await self._display.update_progress_display(
                        self._status.get('title', 'Unknown'),
                        current_time_str,
                        total_time_str,
                        progress
                    )
                
                if progress >= 1:
                    print("✅ Playback complete")
                    if self._use_eink_display:
                        self._display.show_standby()
                    break
            
            await asyncio.sleep(1)

    async def stop(self) -> None:
        """Stop playback"""
        if self._player:
            self._player.stop()
            self._status = {"is_playing": False}
            
            # Update e-ink display to standby mode
            if self._use_eink_display:
                self._display.show_standby()
                
            self._emit('stopped')

    def get_status(self) -> Dict:
        """Get current playback status"""
        return dict(self._status)

    def clear_cache(self) -> None:
        """Clear the audio cache"""
        try:
            for file in self._cache_dir.glob('*'):
                file.unlink()
            print('Cache cleared successfully')
        except Exception as e:
            print(f'Error clearing cache: {e}')

    def __del__(self):
        """Ensure resources are cleaned up"""
        if hasattr(self, '_player') and self._player:
            self._player.stop()
        if hasattr(self, '_vlc_instance') and self._vlc_instance:
            del self._vlc_instance
        
        # Clean up the e-ink display
        if hasattr(self, '_use_eink_display') and self._use_eink_display:
            if hasattr(self, '_display'):
                self._display.cleanup()

    async def __aenter__(self):
        """Support for async context manager"""
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        """Support for async context manager"""
        await self.stop() 